control-plane/api/v1alpha1/servicedefaults_types.go

```diff
@@ -214,6 +218,152 @@ type ServiceDefaultsDestination struct {
+// Rate limiting is a Consul enterprise feature.
+type RateLimits struct {
+	// InstanceLevel represents rate limit configuration
+	// that is applied per service instance.
+	InstanceLevel InstanceLevelRateLimits `json:"instanceLevel,omitempty"`
+}
+
+func (rl *RateLimits) toConsul() *capi.RateLimits {
+	if rl == nil {
+		return nil
+	}
+	routes := make([]capi.InstanceLevelRouteRateLimits, len(rl.InstanceLevel.Routes))
+	for i, r := range rl.InstanceLevel.Routes {
+		routes[i] = capi.InstanceLevelRouteRateLimits{
+			PathExact:         r.PathExact,
+			PathPrefix:        r.PathPrefix,
+			PathRegex:         r.PathRegex,
+			RequestsPerSecond: r.RequestsPerSecond,
+			RequestsMaxBurst:  r.RequestsMaxBurst,
+		}
+	}
+	return &capi.RateLimits{
+		InstanceLevel: capi.InstanceLevelRateLimits{
+			RequestsPerSecond: rl.InstanceLevel.RequestsPerSecond,
+			RequestsMaxBurst:  rl.InstanceLevel.RequestsMaxBurst,
+			Routes:            routes,
+		},
+	}
+}
+
+func (rl *RateLimits) validate(path *field.Path) field.ErrorList {
+	if rl == nil {
+		return nil
+	}
+
+	return rl.InstanceLevel.validate(path.Child("instanceLevel"))
+}
+
+type InstanceLevelRateLimits struct {
+	// RequestsPerSecond is the average number of requests per second that can be
+	// made without being throttled. This field is required if RequestsMaxBurst
+	// is set. The allowed number of requests may exceed RequestsPerSecond up to
+	// the value specified in RequestsMaxBurst.
+	//
+	// Internally, this is the refill rate of the token bucket used for rate limiting.
+	RequestsPerSecond int `json:"requestsPerSecond,omitempty"`
+
+	// RequestsMaxBurst is the maximum number of requests that can be sent
+	// in a burst. Should be equal to or greater than RequestsPerSecond.
+	// If unset, defaults to RequestsPerSecond.
+	//
+	// Internally, this is the maximum size of the token bucket used for rate limiting.
+	RequestsMaxBurst int `json:"requestsMaxBurst,omitempty"`
+
+	// Routes is a list of rate limits applied to specific routes.
+	// For a given request, the first matching route will be applied, if any.
+	// Overrides any top-level configuration.
+	Routes []InstanceLevelRouteRateLimits `json:"routes,omitempty"`
+}
+
+func (irl InstanceLevelRateLimits) validate(path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Track if RequestsPerSecond is set in at least one place in the config
+	isRatelimitSet := irl.RequestsPerSecond > 0
+
+	// Top-level RequestsPerSecond can be 0 (unset) or a positive number.
+	if irl.RequestsPerSecond < 0 {
+		allErrs = append(allErrs,
+			field.Invalid(path.Child("requestsPerSecond"),
+				irl.RequestsPerSecond,
+				"RequestsPerSecond must be positive"))
+	}
+
+	if irl.RequestsPerSecond == 0 && irl.RequestsMaxBurst > 0 {
+		allErrs = append(allErrs,
+			field.Invalid(path.Child("requestsPerSecond"),
+				irl.RequestsPerSecond,
+				"RequestsPerSecond must be greater than 0 if RequestsMaxBurst is set"))
+	}
+
+	if irl.RequestsMaxBurst < 0 {
```

Comment:
error says "RequestsMaxBurst must be greater than 0" should this be `<=0` in the conditional in that case?

Reply:
I'll align it with RequestsPerSecond and say "positive" instead. 0 / unset is a valid value.

Reply:
Oh yes you're right!

Reply:
nevermind, all good, disregard I read that wrong

Reply:
😆